Source/HTTP/Android/http_android.cpp

```diff
@@ -3,13 +3,77 @@
 #include "pch.h"
 #if ANDROID_API
 #include <httpClient/httpClient.h>
+#include "android_http_request.h"
+
+void Internal_HCHttpCallInitialize(void* context)
+{
+    HttpRequest::InitializeJavaEnvironment(reinterpret_cast<JavaVM*>(context));
+}
 
 void Internal_HCHttpCallPerform(
     _In_ AsyncBlock* asyncBlock,
     _In_ hc_call_handle_t call
     )
 {
-    // TODO
+    HC_TRACE_INFORMATION(HTTPCLIENT, "Internal_HCHttpCallPerform");
+    HttpRequest httpRequest;
+    HRESULT result = httpRequest.Initialize();
+
+    if (!SUCCEEDED(result))
+    {
+        CompleteAsync(asyncBlock, result, 0);
+        return;
+    }
+
+    const char* requestUrl = nullptr;
+    const char* requestMethod = nullptr;
+
+    HCHttpCallRequestGetUrl(call, &requestMethod, &requestUrl);
+    httpRequest.SetUrl(requestUrl);
+
+    uint32_t numHeaders = 0;
+    HCHttpCallRequestGetNumHeaders(call, &numHeaders);
+
+    for (uint32_t i = 0; i < numHeaders; i++) 
+    {
+        const char* headerName = nullptr;
+        const char* headerValue = nullptr;
+
+        HCHttpCallRequestGetHeaderAtIndex(call, i, &headerName, &headerValue);
+        httpRequest.AddHeader(headerName, headerValue);
+    }
+
+    const uint8_t* requestBody = nullptr;
+    uint32_t requestBodySize = 0;
+
+    HCHttpCallRequestGetRequestBodyBytes(call, &requestBody, &requestBodySize);
+    if (requestBodySize > 0) 
+    {
+        const char* contentType = nullptr;
+        HCHttpCallRequestGetHeader(call, "Content-Type", &contentType);
+        httpRequest.SetMethodAndBody(requestMethod, contentType, requestBody, requestBodySize);
+    }
+
+    result = httpRequest.ExecuteRequest();
+
+    if (!SUCCEEDED(result)) 
+    { 
+        CompleteAsync(asyncBlock, result, 0);
```

Comment:
>CompleteAsync(asyncBlock, result, 0); [](start = 8, length = 37)

can we set a more detailed error code here?

Reply:
I'm a little confused, should this comment be on ExecuteRequest? We set whatever error code that returns.

That said, it's is returning E_FAIL right now on failure. I think when we look into using a non-blocking call for the HttpRequest we can use that as an opportunity to flesh out some specific error types we'd want call in the async callback.

---
In reply to: [189752347](https://github.com/Microsoft/libHttpClient/pull/165#discussion_r189752347) [](ancestors = 189752347)

Reply:
We can set the platform specific error code on the http request using HCHttpCallResponseSetNetworkErrorCode

---
In reply to: [190067507](https://github.com/Microsoft/libHttpClient/pull/165#discussion_r190067507) [](ancestors = 190067507,189752347)

Reply:
Ah, adding a call to 
HCHttpCallResponseSetNetworkErrorCode(call, result, static_cast<uint32_t>(result)); 
above it.

---
In reply to: [190071569](https://github.com/Microsoft/libHttpClient/pull/165#discussion_r190071569) [](ancestors = 190071569,190067507,189752347)